rust/template/differential_datalog/src/program/mod.rs

```diff
@@ -172,18 +187,258 @@ impl From<i32> for CheckedWeight {
     }
 }
 
+impl From<i8> for CheckedWeight {
+    fn from(item: i8) -> Self {
+        Self { value: item as i32 }
+    }
+}
+
 impl From<CheckedWeight> for i64 {
     fn from(item: CheckedWeight) -> Self {
         item.value as i64
     }
 }
 
+impl From<&CheckedWeight> for i64 {
+    fn from(item: &CheckedWeight) -> Self {
+        item.value as i64
+    }
+}
+
+// Unbounded weights store essentially an unbounded integer, but are
+// optimized for small weights.
+#[derive(Ord, PartialOrd, Eq, PartialEq, Hash, Debug, Clone)]
+pub struct UnboundedWeight {
+    pub value: Either<i32, BigInt>,
+}
+
+impl Semigroup for UnboundedWeight {
+    fn is_zero(&self) -> bool {
+        match &self.value {
+            Left(v) => *v == 0,
+            Right(v) => v.is_zero(),
+        }
+    }
+}
+
+fn to_unbounded_weight(v: BigInt) -> UnboundedWeight {
+    // The invariant is that if the value
+    // fits into an i32 it is stored in the i32, else a BigInt is used.
+    let i = v.to_i32();
+    match i {
+        None => UnboundedWeight { value: Right(v) },
+        Some(r) => UnboundedWeight { value: Left(r) },
+    }
+}
```

Comment:
```suggestion
fn to_unbounded_weight(int: BigInt) -> UnboundedWeight {
    // If the value fits into an i128 it is stored inline, otherwise a BigInt is used.
    let int = int.to_i128();

    match int {
        None => UnboundedWeight::Stack(int),
        Some(int) => UnboundedWeight::Heap(int),
    }
}
```